### ISSUE TITLE
shouldn't hardcode the codepage as 65001 for windows terminal

### ISSUE THREAD

**Issue by ffang** [jline/jline3#164]
As not all windows version can support 65001,  KARAF-5333 is affected by this

https://issues.apache.org/jira/browse/KARAF-5333

**Comment by gnodet** [comment]
This is a regression caused by #133

**Comment by Minecrell** [comment]
@gnodet Well, unfortunately reverting that breaks one part of #133. The reason for hardcoding the UTF-8 code page was that JLine always reads input using UTF-16, but the console is unable to display the characters because Windows defaults to a very naive code page, which barely supports anything.

If I don't change the code page manually before starting the program with this commit, all Cyrillic characters are read properly, but when I type they appear as `????` again in the console. Ideally, requiring users to change the code page manually is something I'd like to avoid.

Maybe there is there a way to check if the code page is supported and only switch in that case?

**Comment by gnodet** [comment]
@ffang and @Minecrell could you have a look at https://github.com/jline/jline3/tree/issue-164 and see if that works for each of you ?

**Comment by ffang** [comment]
@gnodet, no, this not fix the KARAF-5333

**Comment by gnodet** [comment]
@ffang can you try again ?
@Minecrell you'll have to set the {{org.jline.terminal.codepage=65001}} system property

**Comment by Minecrell** [comment]
@gnodet Thanks, this allows me to restore the behaviour of JLine 3.4.0 with the system property so it would do the job.

However, I'm still wondering if there is a way to enable the UTF-8 code page if possible. I'm slightly confused about [KARAF-5333](https://issues.apache.org/jira/browse/KARAF-5333). Is the problem that the UTF-8 code page is not supported by the system (e.g. on older Windows versions)? Or is the problem that the UTF-8 code page doesn't support all characters properly on some Windows version?

If it just doesn't exist, we could use [`IsValidCodePage`](http://go.microsoft.com/fwlink/p/?linkid=178053) to check if it's supported.

**Comment by ffang** [comment]
@gnodet , also doesn't work for me

**Comment by gnodet** [comment]
@ffang the latest commit https://github.com/jline/jline3/commit/7d332545cbe41830561afd5a78da0c594fe088aa should be the same as before, as by default, the code page isn't changed.

**Comment by ffang** [comment]
@gnodet, you are right!
Tested again with latest commit, it works indeed.

The previous error I run into should come from the mistake I made when I resolve the merge conflict. I'm sorry for it.

**Comment by xldai** [comment]
Thanks all! when will be the next version jline 3.5.0 release?
If we want to patch the jline 3.4.0 for this issue, is it enough we cherry pick these 3 commits?
https://github.com/jline/jline3/commit/a9471446325d4290bdf8cf25436ab0ffd5c2da76
https://github.com/jline/jline3/commit/7d332545cbe41830561afd5a78da0c594fe088aa
https://github.com/jline/jline3/commit/cb672b283efd42223ab29511a1ad044ffc606871

**Comment by gnodet** [comment]
I'd like to release it asap, but I need to investigate a few other issues first.
The commits would be 
  https://github.com/jline/jline3/commit/f320221c8b5264cf6e10909a985f57f578b8f564
  https://github.com/jline/jline3/commit/7d332545cbe41830561afd5a78da0c594fe088aa
  https://github.com/jline/jline3/commit/cb672b283efd42223ab29511a1ad044ffc606871

@Minecrell [IsValidCodePage](http://go.microsoft.com/fwlink/p/?linkid=178053) is not supported by jansi, so that can't be done in the short term.  Please raise a new issue if you want to investigate further.

**Comment by Minecrell** [comment]
@gnodet I've been working on a solution that fixes both issues at the same time and avoids all problems caused by the code pages. It's quite different from the current implementation, but so far it is working fine for me. I've opened a pull request at https://github.com/jline/jline3/pull/168

@ffang @xldai Could you test if everything is still working as expected after applying https://github.com/jline/jline3/pull/168?
